src/com/amazon/ionelement/api/StructUpdateBuilder.kt

```diff
@@ -0,0 +1,121 @@
+package com.amazon.ionelement.api
+
+import com.amazon.ionelement.api.StructUpdateBuilder.FieldUpdateMode.*
+import com.amazon.ionelement.impl.*
+import kotlinx.collections.immutable.toPersistentList
+import kotlinx.collections.immutable.toPersistentMap
+
+/**
+ * An [IonPath] represents a traversal into a nested IonElement made of [StructElement]s and [SeqElement]s
+ */
+public class IonPath(public val elements: List<IonPathElement>) : Collection<IonPathElement> by elements {
+    public constructor(vararg elements: IonPathElement) : this(elements.asList())
+
+    public fun tail(): IonPath {
+        return IonPath(elements.drop(1))
+    }
+}
+
+public sealed class IonPathElement
+
+/**
+ * Traverse into a nested [StructElement] by its field name.
+ */
+public data class Field(val fieldName: String) : IonPathElement()
+
+/**
+ * Traverse into a nested [SeqElement] by its numerical index.
+ */
+public data class Index(val index: Int) : IonPathElement()
```

Comment:
> For example, I've tried to consider other possible ways of modeling it, and came up with this.
> ```
> sealed class TreeLocation {
>     data class Path(public val components: List<TreeLocation>): TreeLocation()
>     data class Field(public val fieldName: String): TreeLocation()
>     data class Index(public val index: Int): TreeLocation()
> }
> ```

I was planning on changing it to this pattern because I don't think there is a need to wrap the list of path elements into another class at the moment. One reason to wrap it would be to encapsulate path manipulation functions, but I haven't thought that far ahead yet and they could always be added as extension functions to `List<TreeLocation>`

Reply:
I was not aware of `ion-java-path-extraction` (I had already planned on extending this representation to `sexp`s too). So are you suggesting to allow fetching from an IonElement using a `PathExtractor`? I think that makes sense if there is already a well known way to represent paths.

Reply:
> So are you suggesting to allow fetching from an IonElement using a PathExtractor?

Yes—more or less. It would be nice to see if we can have some similar/familiar APIs across multiple libraries.

I think that the path syntax/model defined in that library could be used to fetch child IonElements from a container. That would fulfill the same use case that you are trying to solve with this PR.

It would also be interesting to be able to extract IonElements directly off of an IonReader (as in `ion-java-path-extraction`), but I am not suggesting that you should do that. (It would be a big scope increase.)

Reply:
...but just to be clear, I don't think `ion-element-kotlin` should take a dependency on `ion-java-path-extraction`. If we want to try to have a unified path extractor API, the classes to model the path should probably be implemented in `ion-java`.